### ISSUE TITLE
CSR_BASE discrepancy

### ISSUE THREAD

**Issue by ozbenh** [enjoy-digital/litex#532]
Commit 748ef1add31bccf
"export: add define of CSR_BASE if not already defined and use it for CSRs definitions/accesses."

Added a CSR_BASE to the generated/csr.h file.

For normal LiteX this is meant to not change anything. The individual CSR accessors
are now expressed in term of CSR_BASE + offset instead of having the origin baked
into the offset but that's about it. At least in theory :)

This feature allows CSR_BASE to be externally redefined however, so that a standalone
core used in a non-LiteX project can have the accessors use whatever origin is appropriate
for that project, based on an address decoder that LiteX doesn't know about. This is how
it is used in Microwatt with standalone LiteDRAM for example.

This is all good except... there is a few of "glitches" in the implementation:

 - It's defined twice ... in generated/csr.h  explicitly (with a ifndef block) and in generated/mem.h implicitly as a result of generating a NAME_BASE for each region NAME (which includes the CSR region). The second definition is implicit.

It should probably be defined only once... Or those two definitions split into two different #define, though as you'll see below, they end up needing to be the same anyways...

 - The two definitions don't necessarily match !

The one in mem.h is the "correct" one imho. It's the origin of the CSR regions in the system bus
(and thus 0 when using a no-core standalone setup) as expected. However we are missing
an ifndef there. 

However the one in csr.h is the origin of the *first* CSR bank in the list of CSR banks. In that code they are called "regions" which confuses things a bit but basically it's not the same as the "CSR  region" in the bus space.

The value *can* be the same if the first CSR bank is bank 0... but there might not be a bank 0 (for example when using a simulated PHY model, litedram bank 0 is empty and goes away).

In that latter case we end up with a CSR_BASE set to, for example, 0x800 in csr.h (and all accessors defined without the 0x800 offset that would have been necessary ssince it's assumed to be included in CSR_BASE). This breaks the intended usage and causes a conflict between the two definitions.

In conclusion, I think the error is in get_csr_header(). csr_base should be passed from the caller
and be the origin of the CSR region in the system bus space, not the offset of whatever CSR bank happens to be first.

In addition, either mem.h should also apply the #ifndef trick or those two definitions should be split using different names for the #define  so that the external definition works properly even if mem.h is included. I got lucky in microwatt (probably due to the order of definitions) but it failed on some test bench I was working with for example.

**Comment by enjoy-digital** [comment]
Fixed with https://github.com/enjoy-digital/litex/commit/c78caeb998156fcb9dc050c60df702cb786e2514.
